pkg/udev/udev.go

```diff
@@ -0,0 +1,132 @@
+// +build linux,cgo
+
+package udev
+
+/*
+  #cgo LDFLAGS: -ludev
+  #include <libudev.h>
+  #include <stdlib.h>
+*/
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/openebs/node-disk-manager/pkg/util"
+)
+
+const (
+	NDMPrefix = "disk-"
+)
+
+func freeCharPtr(s *C.char) {
+	C.free(unsafe.Pointer(s))
+}
+
+// Device wraps a libudev device object
+type Udevice struct {
+	ptr *C.struct_udev_device
+}
+
+// newDevice is a private helper function and returns a pointer to a new device.
+// The device is also added t the devices map in the udev context.
+// The agrument ptr is a pointer to the underlying C udev_device structure.
+// The function returns nil if the pointer passed is NULL.
+func newUdevice(ptr *C.struct_udev_device) (d *Udevice) {
+	// If passed a NULL pointer, return nil
+	if ptr == nil {
+		return nil
+	}
+	// Create a new device object
+	d = &Udevice{
+		ptr: ptr,
+	}
+	// Return the device object
+	return
+}
+
+// PropertyValue retrieves the value of a device property
+func (d *Udevice) PropertyValue(key string) string {
+	k := C.CString(key)
+	defer freeCharPtr(k)
+	return C.GoString(C.udev_device_get_property_value(d.ptr, k))
+}
+
+// SysattrValue retrieves the content of a sys attribute file, and returns an empty string if there is no sys attribute value.
+// The retrieved value is cached in the device.
+// Repeated calls will return the same value and not open the attribute again.
+func (d *Udevice) SysattrValue(sysattr string) string {
+	s := C.CString(sysattr)
+	defer freeCharPtr(s)
+	return C.GoString(C.udev_device_get_sysattr_value(d.ptr, s))
+}
+
+// Devtype returns the devtype string of the udev device.
+func (d *Udevice) Devtype() string {
+	return C.GoString(C.udev_device_get_devtype(d.ptr))
+}
+
+// Devnode returns the device node file name belonging to the udev device.
+// The path is an absolute path, and starts with the device directory.
+func (d *Udevice) Devnode() string {
+	return C.GoString(C.udev_device_get_devnode(d.ptr))
+}
+
+func ListDevices() (m []*Udevice) {
+	udev := C.udev_new()
+	if udev == nil {
+		return nil
+	}
+
+	enumerate := C.udev_enumerate_new(udev)
+	if enumerate == nil {
+		return nil
+	}
+
+	subsystem := C.CString("block")
+	if subsystem == nil {
+		return nil
+	}
+
+	defer freeCharPtr(subsystem)
+
+	err := C.udev_enumerate_add_match_subsystem(enumerate, subsystem)
+	if err < 0 {
+		return nil
+	}
+
+	C.udev_enumerate_scan_devices(enumerate)
+	if err < 0 {
+		return nil
+	}
+
+	m = make([]*Udevice, 0)
+
+	for l := C.udev_enumerate_get_list_entry(enumerate); l != nil; l = C.udev_list_entry_get_next(l) {
+		s := C.udev_list_entry_get_name(l)
+		m = append(m, newUdevice(C.udev_device_new_from_syspath(udev, s)))
+	}
+
+	C.udev_enumerate_unref(enumerate)
```

Comment:
Should this be invoked via `defer` just after enumerate is initialized?

Reply:
changed that.